add top_k parameter to rerank endpoint

aloababa opened the pull request:
This PR adds top_k parameter (number of top documents to return after reranking) to the rerank endpoint.

#342

michaelfeil commented:
Thanks for opening the PR. There need quite some improvements unfortunate.

Align to one of the existing protocols (jina, cohere, voyage, tei), happy to discuss if this is a useful feature.

Expectation would be:
- implementation in "AsyncEngine", and not `infinity_server.py`
- full unit test. 
    - Covering int=0, int=1, int=2, int=very large
    - int=0 is a bad default.
    - Covering from engine side and covering from API side.

aloababa commented:
I've updated the PR with the following changes:

* top_k is now optional and must be greater than 0 on API side.
* top_k is now implemented in AsyncEngine.
* Added tests that covers Engine and API.

michaelfeil commented on libs/infinity_emb/tests/end_to_end/test_torch_reranker.py:
```diff
@@ -89,6 +90,61 @@ async def test_reranker(client, model_base, helpers):
+@pytest.mark.anyio
+async def test_reranker_top_k(client):
+    query = "Where is the Eiffel Tower located?"
+    documents = [
```
Potential flaw: If backend does not do sorting, top_k=1 will just take always the first result.

Better unit test: Wrap this with a unit test, and `return_text`. Make sure that topk=1 solution is always paris in this unit test. e.g. use https://docs.python.org/3/library/itertools.html
````
# someting like:
for return_text in [true, False]
    for raw_score in [True, False]:
          for permutation in itertools.permutation([..paris, ..us, uk]):
               # you test above
```

michaelfeil commented:
@aloababa Thanks for adding all the changes! I am not sure if the top_k is working as intended, I tink it currently just filters the list.
